### ISSUE TITLE
Nested list item rendering layout

### ISSUE THREAD

**Issue by jpolitz** [brownplt/code.pyret.org#67]
When rendering the "items" view of a nested list, CPO produces the following transition:

<img width="363" alt="screen shot 2016-04-22 at 11 02 23 pm" src="https://cloud.githubusercontent.com/assets/119891/14758936/4b8c2840-08de-11e6-9b5c-58f480a24d44.png">

<img width="329" alt="screen shot 2016-04-22 at 11 01 21 pm" src="https://cloud.githubusercontent.com/assets/119891/14758935/3b1e0672-08de-11e6-9275-90c0e51c211d.png">

It would be better if the inner list contents popped down, keeping everything else in a fixed position.

**Comment by sorawee** [comment]
The number of left and right brackets don't even match!

**Comment by sorawee** [comment]
Oh... So the real problem is:

```
       Item 0: 1
[list: Item 1: 2]
```

should actually be:

```
        Item 0: 1
[list:              ]
        Item 1: 2
```

right?

**Comment by blerner** [comment]
No.  The problem with the CSS is there's no good uniform solution,  For lists of images, it looks much better for the pictures to be on a common baseline.  For lists of lists, it looks better for them to have a common first line.  It never works out well for them to be centered, because there's no common sightline to find where anything starts or ends.

**Comment by jpolitz** [comment]
If there's no good uniform solution, we should back up and rethink this, because it's pretty baffling what's going on when these get larger.  Even this case is pretty bad.

Thought: just render constructors so each element in the collection gets its own line.  If there are more than 10 elements in the collection, it starts collapsed.

**Comment by blerner** [comment]
I really don't want to spread everything onto its own line; that looks pretty nasty too -- think of Shriram's spelling-words-by-elements examples, or a simple list of small flags.  There's no reason those have to be line-wrapped.

I didn't spend endless time fussing with the CSS; it's possible that the description I gave above is doable (where a constructor directly inside a constructor gets aligned to the first line, but others get aligned to the baseline, or something like that).

**Comment by blerner** [comment]
One alternative possibility is that on mouse-hovering within the rendering of an expression, we shade the background slightly, so that it becomes more obvious what the extent of that value is, without needing annoying permanent outlines around everything.

**Comment by jpolitz** [comment]
I don't see why flags or the spelling words would be "nasty" on separate lines.  Really, I don't.

Maybe just for _nested_ collections the one-per-line thing makes sense.

Now that I think about it, I would be a little confused by the list example whether it popped up or down, because if the rendering happens to span multiple lines, then things would jump around as well (since the width would change).  I really think a more structured layout would make this more predictable and useful.

**Comment by blerner** [comment]
Spelling words is easy: English words aren't spelled one-or-two letters per line.  Part of the appeal of that rendering is the words simply read across.  Typographically, humans are good at reading about 70 characters per line, and artificially breaking that up just looks clunky.

On the other hand, having things line-wrap smoothly means that they'll move around when we change layouts.  This is already trivially true if you toggle a fraction from `1/7` to `0.142857` and those extra characters cause new line breaks.

I don't know what you mean by a "more structured layout" here -- we want our primary layout to look like constructor notation, i.e. linear, textual syntax, with embedded pretty pictures and tables (which, btw, will completely and totally break any nice rendering we want, because now we also need to constrain cells to readable widths so the columns don't overflow).    You can't have it both ways here, and I think the far lesser evil is for things to move around when you explicitly click on something to change its rendering.

**Comment by jpolitz** [comment]
By "more structured layout" I mean something like what the Chrome console does.

I'm good at reading 70 characters worth of code or prose per line, but I'm not good at reading 70 character-wrapped constructor printing with no newlines or indentation.  Maybe other humans are.

Here's a flow that illustrates what I mean (this was the first thing I tried, I'm not fishing for bad examples):

<img width="649" alt="screen shot 2016-05-11 at 12 02 25 pm" src="https://cloud.githubusercontent.com/assets/119891/15187782/83ede9a8-1770-11e6-9caf-0bf5395f656c.png">

Then I click one of the objects that is left aligned.  It shrinks in width and pops up a line:

<img width="643" alt="screen shot 2016-05-11 at 12 02 56 pm" src="https://cloud.githubusercontent.com/assets/119891/15187831/b69dca12-1770-11e6-87da-88817cab250c.png">

Then I close the top one, and the bottom one moves to yet another spot:

<img width="643" alt="screen shot 2016-05-11 at 12 02 56 pm" src="https://cloud.githubusercontent.com/assets/119891/15187848/ca7370aa-1770-11e6-8fbb-bbcf07443152.png">

Also, my cursor isn't hovering over the value it was just over, so my place in the list is gone.

This rendering and adjusting is really annoying as a way of exploring.

Now, I know that I can click on the "list" view and get the row-per-item view, which is really nice.  But the stuff going on when I click the inner object isn't very helpful at all.

Maybe the whole thing should pop open if it's a collection and we click on some inner item, maybe something else.  But I think we can agree that the above is kind of baffling to work with?

**Comment by blerner** [comment]
Yes that's baffling, but no, I don't think the Chrome console approach is better *as a default* for users -- in fact, it's exactly as bad, if not worse.  Here's your example done in JS in the Chrome console (for an apples-to-apples comparison):
![ss1](https://cloud.githubusercontent.com/assets/918464/15188147/4b157a18-1772-11e6-815e-e149bfc078d1.png)
And if I click on one of those objects (I have no idea what I'm looking at, because there's zero information visible, so let's click on the one in the first row on the right):
![ss2](https://cloud.githubusercontent.com/assets/918464/15188194/79bf95ec-1772-11e6-9586-ed36548304e1.png)
Sure enough, it gets wider and drops down a line, too, and my cursor isn't over it, either.  Same objection as yours, just triggered moving downward instead of upward.  Likewise if I expand a second object later in the list, and then collapse the earlier one, the second one jumps around, too.

Maybe the solution here is you can only expand the outermost collection.  That's really annoying if the thing you want is deeply nested and you don't want to expand all the surrounding stuff.

Maybe the solution here is to expand everything outward from whatever you wanted to expand.  But that's even more jarring and moves even more things around, and you've completely lost any sense of where your relevant data has gone.

Any solution that involves line wrapping will involve rewrapping.  And any solution that involves hard linebreaks will be suboptimal in the common cases...

**Comment by jpolitz** [comment]
Huh, thanks for pointing that out about Chrome console.  I was so used to their object pop-down, and what happens for lists of more than 100 elements, that I thought it was the default.  Turns out Pyret's equally good/bad for the cases I'm thinking of.  Pyret's potentially better on shallow nesting, and potentially more confusing on deep nesting (at some point values should be truncated).

I think I don't have this clearly down to one issue yet, and tried to broaden it too much in the ensuing discussion.

Going back to just "it should pop down, not up, on nested constructors" might be the best concrete thing to suggest, but I doubt that covers all the cases.

**Comment by shriram** [comment]
For now, can we just use top-alignment? Bottom-alignment _may_ be a bit better looking for image cases, but the output is so confusing to a reader in the bad cases that “ugliness” is worth the sacrifice? Let's see whether it gets worse than ugly.
